### PR TITLE
FIX: Rewrapped ?gels and ?gels_lwork routines

### LINKED ISSUE #8329: `_gels` fails for fat A matrix
Replicating the tests in `scipy/linalg/tests/test_lapack.py` for a fat A matrix, `dgels` fails with `ValueError: Internal work array size computation failed: -8`. 

```python
m = 10
n = 20

a1 = np.ones((m, n))
b1 = np.ones(m)
gels, gels_lwork, geqrf = get_lapack_funcs(
        ('gels', 'gels_lwork', 'geqrf'), (a1, b1))

m, n = a1.shape
if len(b1.shape) == 2:
    nrhs = b1.shape[1]
else:
    nrhs = 1

# Request of sizes
lwork = _compute_lwork(gels_lwork, m, n, nrhs)
```

This is probably because LAPACK [asks](http://www.netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_ga225c8efde208eaf246882df48e590eac.html) for `LDB >= MAX(1,M,N)`, while this line 
https://github.com/scipy/scipy/blob/ad4a557eadbc5d940f33f00a9dd9729ad7184fd3/scipy/linalg/flapack.pyf.src#L1481
sets ldb to M, which might be less than N.

### PR DESCRIPTION
**Pull request by ilayn**
Closes #8329 

Regrouped the declarations under one subroutine and added more safety nets for the array size declarations. Simple test added.

### CODE REVIEW

**Comment by ilayn** [comment]
Anybody can replicate these seemingly segfaults? I can't on my Win and Linux machines.

**Comment by pstjohn** [comment]
So it runs without a segfault for me. But not sure its giving the right answer: we expect `x` to be length `n`, while gels still returns one of length `m`.

Perhaps a better test case would be 

```python
a = np.array([[1, 1, 1], [-1, -1, 1]])
b = np.array([1, 0])

m, n = a.shape
rhs = 1

lwork = _compute_lwork(gels_lwork, m, n, nrhs)
lqr, x, info = gels(a, b, lwork=lwork)

assert np.allclose(x, np.array([0.25, 0.25, 0.5]))
```

**Comment by ilayn** [comment]
I think b vector must be at least `max(m,n)` regardless of the solution length. That's probably causing the segfaults when LAPACK tries to access the `[m+1,n]` parts of the array.

**Comment by pstjohn** [comment]
exactly, they do that when calling the various routines. https://github.com/scipy/scipy/blob/7fadcdcd7995ab581fdaaf68cb2b7199c7685a01/scipy/linalg/basic.py#L1184-L1186

But with `_gels` I get an error when passing a differently sized b matrix

File: scipy/linalg/flapack.pyf.src [hunk 1]
```diff
@@ -1464,102 +1464,50 @@ interface
...
-     integer optional,intent(in),depend(nrhs,m,n),&
-          check(lwork>=1||lwork==-1) &
-          :: lwork=MAX( MIN(m,n) + MAX(MIN(m,n), nrhs ), 1 )
-     <ftype2> dimension(m,nrhs),intent(in,out,copy,out=x) :: b
-     <ftype2> dimension(m,n),intent(in,out,copy,out=lqr) :: a
-     <ftype2> depend(lwork),dimension(MAX(lwork,1)),intent(hide,cache):: work
+     integer intent(hide),depend(b):: nrhs = shape(b,1)
+     <ftype> dimension(m,n),intent(in,out,copy,out=lqr):: a
+     integer intent(hide),depend(a):: lda = MAX(1,shape(a,0))
+     <ftype> intent(in,out,copy,out=x),depend(trans,m,n),dimension((*trans=='N'?m:n),nrhs),check(shape(b,0)==(*trans=='N'?m:n)) :: b
```
Comment (pstjohn) [inline scipy/linalg/flapack.pyf.src]: what does `trans=='N'?m:n` do? I'm assuming though that we'll need some logic with `MAX(m,n)`, if you look at the gelss routines that's how they handle the b matrix

Comment (ilayn) [reply scipy/linalg/flapack.pyf.src]: Yes I probably overlooked the extended b vector part. The routine works when that part is fixed. I'll just sweep one more time for the corner cases and update.

**Comment by ilayn** [comment]
I think this time it is working.

**Comment by pstjohn** [comment]
woohoo! thanks for looking into that! This looks great.

**Comment by ilayn** [comment]
In case noone else wants review this, I am +1 to push this in a few days.

**Review by rgommers — APPROVED** [review]

**Pull request merged**